blas/impl/KokkosBlas2_ger_impl.hpp

```diff
@@ -0,0 +1,339 @@
+  AlphaCoeffType                 alpha_;
+  typename XViewType::const_type x_;
+  typename YViewType::const_type y_;
+  AViewType                      A_;
+};
+
+// Single-level parallel version of GER.
+template <class XViewType, class YViewType, class AViewType,
+          class IndexType = typename AViewType::size_type>
+void singleLevelGer( const typename AViewType::execution_space  & space
+                   , const          char                          trans[]
+                   , const typename AViewType::const_value_type & alpha
+                   , const          XViewType                   & x
+                   , const          YViewType                   & y
+                   , const          AViewType                   & A
+                   ) {
+  KOKKOS_IMPL_DO_NOT_USE_PRINTF( "Entering IMPL singleLevelGer(), AViewType = %s\n", typeid(AViewType).name() );
+  static_assert(Kokkos::is_view<XViewType>::value, "XViewType must be a Kokkos::View.");
+  static_assert(Kokkos::is_view<YViewType>::value, "YViewType must be a Kokkos::View.");
+  static_assert(Kokkos::is_view<AViewType>::value, "AViewType must be a Kokkos::View.");
+
+  static_assert(static_cast<int>(XViewType::rank) == 1, "XViewType must have rank 1.");
+  static_assert(static_cast<int>(YViewType::rank) == 1, "YViewType must have rank 1.");
+  static_assert(static_cast<int>(AViewType::rank) == 2, "AViewType must have rank 2.");
+
+  static_assert(std::is_integral<IndexType>::value, "IndexType must be an integer");
+
+  using KAT = Kokkos::ArithTraits<typename AViewType::non_const_value_type>;
+
+  if (y.extent(0) == 0) {
+    // no entries to update
+  }
+  else if (x.extent(0) == 0) {
+    // no entries to update
+  }
+  else if (alpha == KAT::zero()) {
+    // no entries to update
+  }
+  else {
+    using execution_space = typename AViewType::execution_space;
+    Kokkos::RangePolicy<execution_space, IndexType>            rangePolicy(space, 0, A.extent(0));
+    SingleLevelGER<XViewType, YViewType, AViewType, IndexType> functor( (trans[0] == 'T') || (trans[0] == 't')
+                                                                      , alpha
+                                                                      , x
+                                                                      , y
+                                                                      , A
+                                                                      );
+    Kokkos::parallel_for("KokkosBlas::ger[SingleLevel]", rangePolicy, functor);
+  }
+}
+
+struct TwoLevelGER_LayoutLeftTag {};
+struct TwoLevelGER_LayoutRightTag {};
+
+// ---------------------------------------------------------------------------------------------
+
+// Functor for a two-level parallel_reduce version of GER, designed for performance on GPU.
+// Kernel depends on the layout of A.
+template <class XViewType, class YViewType, class AViewType, class IndexType = typename AViewType::size_type>
+struct TwoLevelGER {
```

Comment:
@lucbv we should get rid of the `TwoLevel` naming and have consistent ones that follows: https://github.com/kokkos/kokkos-kernels/wiki/Batched-BLAS-and-LAPACK 
It looks like only GEMV currently uses `TwoLevel`. I would be in favor of standardizing the names.

Reply:
I am not sure what naming scheme you are referring?
Also note that these are internal implementations so we are even less likely to have a naming scheme for that?

Reply:
Actually, forget this comment, this is not a TeamVector kernel.